pkg/sfu/buffer/frameintegrity.go

```diff
@@ -0,0 +1,211 @@
+		return
+	}
+
+	if fe.startSeq == nil && ddVal.FirstPacketInFrame {
+		fe.startSeq = &extSeq
+	}
+	if fe.endSeq == nil && ddVal.LastPacketInFrame {
+		fe.endSeq = &extSeq
+	}
+
+	if fe.startSeq != nil && fe.endSeq != nil {
+		if fe.packetsConsective(*fe.startSeq, *fe.endSeq) {
+			fe.integrity = true
+		}
+	}
+}
+
+func (fe *FrameEntity) Reset() {
+	fe.integrity = false
+	fe.startSeq, fe.endSeq = nil, nil
+}
+
+func (fe *FrameEntity) Integrity() bool {
+	return fe.integrity
+}
+
+// ------------------------------
+
+type PacketHistory struct {
+	base        uint64
+	last        uint64
+	bits        []uint64
+	packetCount int
+	inited      bool
+}
+
+func NewPacketHistory(packetCount int) *PacketHistory {
+	packetCount = (packetCount + 63) / 64 * 64
+	return &PacketHistory{
+		bits:        make([]uint64, packetCount/64),
+		packetCount: packetCount,
+	}
+}
+
+func (ph *PacketHistory) AddPacket(extSeq uint64) {
+	if !ph.inited {
+		ph.inited = true
+		ph.base = uint64(extSeq)
+		// set base to extSeq-100 to avoid out-of-order packets belongs to first frame to be dropped
+		if ph.base > 100 {
+			ph.base -= 100
+		} else {
+			ph.base = 0
+		}
+		ph.last = uint64(extSeq)
+		ph.set(extSeq, true)
+		return
+	}
+
+	if extSeq <= ph.base {
+		// too old
+		return
+	}
+
+	if extSeq <= ph.last {
+		if ph.last-extSeq < uint64(ph.packetCount) {
+			ph.set(extSeq, true)
+		}
+		return
+	}
+
+	for i := ph.last + 1; i < extSeq; i++ {
+		ph.set(i, false)
+	}
+
+	ph.set(extSeq, true)
+	ph.last = extSeq
+}
+
+func (ph *PacketHistory) getPos(seq uint64) (index, offset int) {
+	idx := (seq - ph.base) % uint64(ph.packetCount)
+	return int(idx >> 6), int(idx % 64)
+}
+
+func (ph *PacketHistory) set(seq uint64, received bool) {
+	idx, offset := ph.getPos(seq)
+	if !received {
+		ph.bits[idx] &= ^(1 << offset)
+	} else {
+		ph.bits[idx] |= 1 << (offset)
+	}
+}
+
+func (ph *PacketHistory) PacketsConsective(start, end uint64) bool {
+	if start > end {
+		return false
+	}
+
+	if end-start >= uint64(ph.packetCount) {
+		return false
+	}
+
+	startIndex, startOffset := ph.getPos(start)
+	endIndex, endOffset := ph.getPos(end)
+
+	if startIndex == endIndex && end-start <= 64 {
+		testBits := uint64((1<<(endOffset-startOffset+1))-1) << startOffset
+		return ph.bits[startIndex]&testBits == testBits
+	}
+
+	if (ph.bits[startIndex]>>(startOffset))+1 != 1<<(64-startOffset) {
+		return false
+	}
+
+	for i := startIndex + 1; i != endIndex; i++ {
+		if i == len(ph.bits) {
+			i = 0
+			if i == endIndex {
+				break
+			}
+		}
+		if ph.bits[i]+1 != 0 {
+			return false
+		}
+	}
+
+	testBits := uint64((1 << (endOffset + 1)) - 1)
+	return ph.bits[endIndex]&testBits == testBits
+}
+
+// ------------------------------
+
+type FrameIntegrityChecker struct {
+	frameCount int
+	frames     []FrameEntity
+	base       uint64
+	last       uint64
+
+	pktHistory *PacketHistory
+	inited     bool
+}
+
+func NewFrameIntegrityChecker(frameCount, packetCount int) *FrameIntegrityChecker {
```

Comment:
There is a frame chain in video layer selector on subscriber too. That is also checking for broken chains. Is it possible to combine the two? Mainly do the work here and tell subscribers. Each subscriber doing the work adds up.

Reply:
It is a frame integrity checker here to trace a single frame that split to packets, not tracing the frame chain. Have thought of merging them in one place but can't do that because each subscriber has its own decode target and forwarder state, so one frame chain might be broken in subscriber A but kept intact in subscriber B (because they are forwarding different frames). But a potential improvement is detecting the frame chain both in upstream and downstream, it will not eliminate the frame chain tracing in the downstream but can speed up the chain broken detection as downstream's chain is subject to the upstream's chain.

Reply:
Got it. Thank you @cnderrauber .